sdks/python/apache_beam/task/task_worker/k8s/handler.py

```diff
@@ -0,0 +1,221 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Kubernetes task worker implementation.
+"""
+
+from __future__ import absolute_import
+
+import threading
+import time
+from typing import TYPE_CHECKING
+
+from apache_beam.runners.worker.task_worker.handlers import TaskWorkerHandler
+
+try:
+  from kubernetes.client import BatchV1Api
+  from kubernetes.client import V1EnvVar
+  from kubernetes.client import V1Container
+  from kubernetes.client import V1PodTemplateSpec
+  from kubernetes.client import V1PodSpec
+  from kubernetes.client import V1ObjectMeta
+  from kubernetes.client import V1Job
+  from kubernetes.client import V1JobSpec
+  from kubernetes.client import V1DeleteOptions
+  from kubernetes.client.rest import ApiException
+  from kubernetes.config import load_kube_config
+except ImportError:
+  BatchV1Api = None
+  V1EnvVar = None
+  V1Container = None
+  V1PodTemplateSpec = None
+  V1PodSpec = None
+  V1ObjectMeta = None
+  V1Job = None
+  V1JobSpec = None
+  V1DeleteOptions = None
+  ApiException = None
+  load_kube_config = None
+
+if TYPE_CHECKING:
+  from typing import List
+
+__all__ = [
+    'KubeTaskProperties',
+    'KubeTaskWorkerHandler',
+]
+
+
+class KubeTaskProperties(object):
+  """
+  Object for describing kubernetes job properties for a task worker.
+  """
```

Comment:
I expected this object essentially be a the `spec` portion of the kube `Job` object, as a dict.  

let's consider this example job definition from the docs:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: pi
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```

Which keys does the `KubeJobManager` override or set defaults for?   Presumably everything else could be something that a user might want to provide.

Reply:
It can be used to pass along values for any field in the construction of the kubernetes job. It's used primarily within `KubeTaskWorkerHandler.create_job` but is very minimal at the moment. We should grow support for specifying the `ResourceQuota`  along with specifying [how these execute in parallel](https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs).

Currently all defaults are just defined on `KubeTaskProperties.__init__`.

The `KubeJobManager` is just a thread for monitoring the k8s jobs so if they get deleted we can notify the handler.

I'm thinking that the `KubeTask` payload should simply be a `kubernetes.client.V1Job` object directly rather than attempting to abstract everything. What are your thoughts on that?

Reply:
> I'm thinking that the KubeTask payload should simply be a `kubernetes.client.V1Job` object directly rather than attempting to abstract everything. What are your thoughts on that?

Yes, exactly this.  Would `kubernetes.client.V1Job` replace `KubeTaskProperties` or be a member of `KubeTaskProperties`?

Reply:
I think I would ditch the `KubeTaskProperties` all together. The only thing that's a bit awkward is we need to patch in some env vars into the job within the handler before we submit.

Reply:
I ended up keeping a light wrapper object `KubePayload` that contains a namespace (str) and a job (V1Job). The job object unfortunately doesn't have any concept of namespace so I needed to pass that along somehow. The `KubeTask` is now responsible for constructing the job object instead of that happening within the handler - which enables the wrapper (per-task tweak of the payload) to have access to the job directly.